accepted/future-releases/variance/feature-specification.md

```diff
@@ -0,0 +1,373 @@
+the type alias.*
+
+We say that a type parameter _X_ of a type alias _F_ _is covariant_ if it
+only occurs covariantly in the body of _F_; that it _is contravariant_ if
+it occurs contravariantly in the body of _F_ and does not occur covariantly
+or invariantly; that it _is invariant_ if it occurs invariantly in the body
+of _F_ (*with no constraints on other occurrences*), or if it occurs both
+covariantly and contravariantly.
+
+*In particular, an unused type parameter is considered covariant.*
+
+Let _D_ be the declaration of a class or mixin, and let _X_ be a type
+parameter declared by _D_.
+
+If _X_ has the variance modifier `out` then it is a compile-time error for
+_X_ to occur in a non-covariant position in a member signature in the body
+of _D_, except that it is not an error if it occurs in a covariant position
+in the type annotation of a covariant formal parameter (*this is a
+contravariant position in the member signature as a whole*).
+
+*In particular, _X_ can not be the type of a method parameter (unless
+covariant), and it can not be the bound of a type parameter of a generic
+method.*
+
+If _X_ has the variance modifier `in` then it is a compile-time error for
+_X_ to occur in a non-contravariant position in a member signature in the
+body of _D_, except that it is not an error if it occurs in a contravariant
+position in the type of a covariant formal parameter. *For instance, _X_
+can not be the return type of a method or getter, and it can not be the
+bound of a type parameter of a generic method.*
+
+*If _X_ has the variance modifier `inout` then there are no variance
+related restrictions on the positions where it can occur in member
+signatures.*
+
+Let _D_ be a class or mixin declaration, let _S_ be a direct superinterface
+of _D_, and let _X_ be a type parameter declared by _D_.  It is a
+compile-time error if _X_ is covariant and _X_ occurs in a non-covariant
+position in _S_. It is a compile-time error if _X_ is contravariant, and
+_X_ occurs in a non-contravariant position in _S_.  In these rules, type
+inference of _S_ is assumed to have taken place already.
+
+*An invariant type parameter can occur in any position in a superinterface.
+These constraints on allowed locations for type parameters ensure that if
+we consider type arguments _Args1_ and _Args2_ passed to _D_ such that the
+former produces a subtype, then we also have _S1 <: S2_ where _S1_ and _S2_
+are the corresponding instantiations of _S_.*
+
+```dart
+class A<out X, inout Y, in Z> {}
+class B<out U, inout V, in W> implements
+    A<U Function(W), V Function(V), W Function(V)> {}
+
+// B<int, String, num> <: B<num, String, int>, and hence
+// A<int Function(num), String Function(String), num Function(String)> <:
+// A<num Function(int), String Function(String), int Function(String)>.
+```
+
+*In a superinterface, a type parameter without a variance modifier can be
+used in an actual type argument for a parameter with a variance modifier,
```

Comment:
Worth making it explicit that this is subject to the checks above, which de facto means that the vice versa only allows the "explicitly covariant or invariant to implicitly covariant" cases?  That is, `class C<in X> implements List<X>` is a no-no, right?

And don't we have the "twisted super-hierarchy" problem to deal with here too, so we have to outlaw "class C<X> implements Contra<X>, Invariant<X>"?

Reply:
> And don't we have the "twisted super-hierarchy" problem to deal with here too, so we have to outlaw "class C implements Contra, Invariant"?

On further thought, this is I think disallowed by our existing super-interface check, but probably worth pointing that out here.

Reply:
I added a separate section about the dynamic checks.